subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java

```diff
@@ -48,35 +80,40 @@ public boolean containsValue(Object value) {
 
     @Override
     public Object get(Object key) {
+        if (key instanceof String) {
+            return attributes.get(keyFor((String) key));
+        }
         return attributes.get(key);
     }
 
     @Override
     public Object put(String key, Object value) {
         if (key == null) {
-            throw new ManifestException("The key of a manifest attribute must not be null.");
+            // NPE is the specified exception type (see Map#put javadoc)
+            throw new NullPointerException("The key of a manifest attribute must not be null.");
         }
         if (value == null) {
-            throw new ManifestException(String.format("The value of a manifest attribute must not be null (Key=%s).", key));
+            // NPE is the specified exception type (see Map#put javadoc)
+            throw new NullPointerException(String.format("The value of a manifest attribute must not be null (Key=%s).", key));
+        }
+        Name name;
+        if (key.length() > 70 || key.length() * 3 > 70 && Utf8.encodedLength(key) > 70) {
+            throw new ManifestException(String.format("The Key=%s violates the Manifest spec: key should not exceed 70 bytes, actual length is %d bytes in UTF-8", key, Utf8.encodedLength(key)));
```

Comment:
❓  This may be a breaking change. What is the current behaviour? If it's obviously broken when a user doesn't respect those limits (fails at a later stage), then we could start throwing an exception. In the case where nothing obvious breaks, then we should use a `DeprecationLogger` to nag about this becoming an error in the next Gradle major version.

Reply:
Frankly speaking, in 99.42% of the cases, everybody uses `plain old ascii` for **keys**.
Thus string length and UTF-8 length are basically the same.
However, the manifest spec forbids to split manifest keys, so manifest parsers might easily fail to parse those split key names.

Even though it is a breaking change, I don't think it is really important to somehow support long manifest keys.


However, if you feel strongly I can add `DeprecationLogger` (or whatever), however, I would need a pointer to the similar feature. Gradle's codebase is huge, and I'm not familiar with approaches to "feature flags", "deprecation warnings", etc, etc.